### ISSUE TITLE
error while building seiri-watcher

### ISSUE THREAD

**Issue by cv127001** [chyyran/seiri#15]
Hello, 

I am following build.md and after installing Rust, CMake, and Node successfully, I attempted to build "seiri-watcher" and got the following error:

```
carlos@dp:~/dev-rust/seiri/seiri-watcher$ cargo build --release
    Updating crates.io index
error: failed to get `notify` as a dependency of package `seiri-watcher v0.2.0 (/home/carlos/dev-rust/seiri/seiri-watcher)`

Caused by:
  failed to load source for dependency `notify`

Caused by:
  Unable to update /home/carlos/dev-rust/seiri/seiri-watcher/notify

Caused by:
  failed to read `/home/carlos/dev-rust/seiri/seiri-watcher/notify/Cargo.toml`

Caused by:
  No such file or directory (os error 2)
```

My environment is the following:
```
carlos@dp:~/dev-rust/seiri/seiri-watcher$ uname -a
Linux dp 5.8.0-7630-generic #32~1607010078~20.04~383a644-Ubuntu SMP Thu Dec 3 19:14:47 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

```

I'm anxious to use this. Please respond at your earliest convenience. Thank you! 

Carlos

**Comment by chyyran** [comment]
Hi Carlos,

It's probably due for an update but `seiri` uses a forked version of `notify-rs` because 4.0 wasn't stable when I was updating this.

You can fix that by initializing the submodules

```bash
git submodule update --init --recursive
```

Then it should work.

**Comment by cv127001** [comment]
Ronny,

Thank you **very much** for the quick reply.

Your suggestion helped, so thank you.

I ran it but have a new hurdle:

`carlos@dp:~/dev-rust/seiri$ git submodule update --init --recursive
Submodule 'seiri-watcher/notify' (https://github.com/chyyran/notify) registered for path 'seiri-watcher/notify'
Cloning into '/home/carlos/dev-rust/seiri/seiri-watcher/notify'...
Submodule path 'seiri-watcher/notify': checked out '5150ed3f091f51498fa892892158482da30aa7e3'
carlos@dp:~/dev-rust/seiri$ ls
branding  Cargo.lock  LICENSE.md  seiri-client            seiri-lib   seiri-watcher  yarn-error.log
build.md  katatsuki   README.md   seiri-client-internals  seiri-neon  testtracks.db
carlos@dp:~/dev-rust/seiri$ cd seiri-watcher/
carlos@dp:~/dev-rust/seiri/seiri-watcher$ cargo build --release
  Downloaded semver v0.9.0
  Downloaded semver-parser v0.7.0
  Downloaded rustc_version v0.2.3
  Downloaded nom v3.2.1
  Downloaded syn v0.11.11
  Downloaded lazy_static v0.2.11
  Downloaded glob v0.2.11
  Downloaded rand_core v0.2.2
  Downloaded quote v0.5.2
  Downloaded parking_lot v0.6.4
  Downloaded bitflags v0.9.1
  Downloaded syn v0.13.11
  Downloaded vec_map v0.8.1
  Downloaded unreachable v1.0.0
  Downloaded quick-error v1.2.1
  Downloaded crossbeam-epoch v0.6.1
  Downloaded ansi_term v0.11.0
  Downloaded num-integer v0.1.38
  Downloaded chrono v0.4.2
  Downloaded unicode-xid v0.1.0
  Downloaded bitflags v1.0.4
  Downloaded arrayvec v0.4.7
  Downloaded xdg v2.1.0
  Downloaded iovec v0.1.2
  Downloaded antidote v1.0.0
  Downloaded memchr v1.0.2
  Downloaded num-traits v0.1.43
  Downloaded log v0.3.9
  Downloaded aster v0.41.0
  Downloaded atty v0.2.11
  Downloaded stable_deref_trait v1.1.1
  Downloaded scopeguard v0.3.3
  Downloaded regex v0.2.11
  Downloaded quasi v0.32.0
  Downloaded regex-syntax v0.5.6
  Downloaded threadpool v1.7.1
  Downloaded quote v0.3.15
  Downloaded proc-macro2 v0.3.8
  Downloaded humantime v1.1.1
  Downloaded parking_lot_core v0.3.1
  Downloaded tokio-executor v0.1.5
  Downloaded mio v0.6.16
  Downloaded void v1.0.2
  Downloaded term v0.4.6
  Downloaded syntex_errors v0.58.1
  Downloaded ucd-util v0.1.1
  Downloaded unicode-xid v0.0.4
  Downloaded textwrap v0.10.0
  Downloaded mio-extras v2.0.5
  Downloaded slab v0.4.1
  Downloaded rand v0.5.5
  Downloaded toml v0.4.6
  Downloaded thread_local v0.3.5
  Downloaded utf8-ranges v1.0.0
  Downloaded cfg-if v0.1.3
  Downloaded aho-corasick v0.6.4
  Downloaded dirs v1.0.3
  Downloaded rustc-serialize v0.3.24
  Downloaded inotify-sys v0.1.3
  Downloaded filetime v0.2.3
  Downloaded enum-primitive-derive v0.1.2
  Downloaded clang-sys v0.19.0
  Downloaded app_dirs v1.2.1
  Downloaded time v0.1.40
  Downloaded itertools v0.7.8
  Downloaded tokio-reactor v0.1.6
  Downloaded libkatatsuki-sys v1.0.9
  Downloaded leak v0.1.2
  Downloaded bitflags v0.8.2
  Downloaded lazy_static v1.2.0
  Downloaded crossbeam v0.5.0
  Downloaded imagesize v0.5.2
  Downloaded crossbeam-utils v0.5.0
  Downloaded rand v0.4.2
  Downloaded syntex v0.58.1
  Downloaded rand_core v0.3.0
  Downloaded serde v1.0.53
  Downloaded memchr v2.0.1
  Downloaded nodrop v0.1.13
  Downloaded num_cpus v1.8.0
  Downloaded peeking_take_while v0.1.2
  Downloaded net2 v0.2.32
  Downloaded unicode-width v0.1.5
  Downloaded owning_ref v0.3.3
  Downloaded katatsuki v1.0.9
  Downloaded pkg-config v0.3.11
  Downloaded quasi_codegen v0.32.0
  Downloaded synom v0.11.3
  Downloaded rusqlite v0.15.0
  Downloaded r2d2_sqlite v0.7.0
  Downloaded lazycell v1.2.0
  Downloaded cexpr v0.2.3
  Downloaded syntex_syntax v0.58.1
  Downloaded walkdir v2.1.4
  Downloaded libsqlite3-sys v0.10.0
  Downloaded memoffset v0.2.1
  Downloaded tokio-io v0.1.10
  Downloaded crossbeam-utils v0.6.1
  Downloaded bindgen v0.26.3
  Downloaded smallvec v0.6.5
  Downloaded serde_derive v1.0.53
  Downloaded linked-hash-map v0.4.2
  Downloaded crossbeam-deque v0.6.2
  Downloaded bytes v0.4.10
  Downloaded libloading v0.4.3
  Downloaded byteorder v1.2.7
  Downloaded env_logger v0.4.3
  Downloaded scheduled-thread-pool v0.2.0
  Downloaded same-file v1.0.2
  Downloaded lock_api v0.1.4
  Downloaded crossbeam-channel v0.3.1
  Downloaded inotify v0.6.1
  Downloaded num-traits v0.2.4
  Downloaded cc v1.0.15
  Downloaded syntex_pos v0.58.1
  Downloaded log v0.4.1
  Downloaded clap v2.32.0
  Downloaded cmake v0.1.33
  Downloaded lru-cache v0.1.1
  Downloaded either v1.5.0
  Downloaded libc v0.2.43
  Downloaded r2d2 v0.8.3
  Downloaded futures v0.1.25
  Downloaded strsim v0.7.0
  Downloaded 124 crates (7.6 MB) in 1.73s (largest was `libsqlite3-sys` at 2.2 MB)
   Compiling libc v0.2.43
   Compiling unicode-xid v0.0.4
   Compiling cfg-if v0.1.3
   Compiling rustc-serialize v0.3.24
   Compiling term v0.4.6
   Compiling bitflags v0.8.2
   Compiling void v1.0.2
   Compiling cc v1.0.15
   Compiling regex v0.2.11
   Compiling num-traits v0.2.4
   Compiling lazy_static v1.2.0
   Compiling semver-parser v0.7.0
   Compiling rand_core v0.3.0
   Compiling glob v0.2.11
   Compiling ucd-util v0.1.1
   Compiling libloading v0.4.3
   Compiling scopeguard v0.3.3
   Compiling unicode-width v0.1.5
   Compiling utf8-ranges v1.0.0
   Compiling stable_deref_trait v1.1.1
   Compiling strsim v0.7.0
   Compiling bitflags v1.0.4
   Compiling bitflags v0.9.1
   Compiling vec_map v0.8.1
   Compiling ansi_term v0.11.0
   Compiling nodrop v0.1.13
   Compiling lazy_static v0.2.11
   Compiling pkg-config v0.3.11
   Compiling peeking_take_while v0.1.2
   Compiling memoffset v0.2.1
   Compiling num-integer v0.1.38
   Compiling byteorder v1.2.7
   Compiling futures v0.1.25
   Compiling slab v0.4.1
   Compiling lazycell v1.2.0
   Compiling unicode-xid v0.1.0
   Compiling linked-hash-map v0.4.2
   Compiling antidote v1.0.0
   Compiling quote v0.3.15
   Compiling crossbeam-utils v0.5.0
   Compiling quick-error v1.2.1
   Compiling same-file v1.0.2
   Compiling serde v1.0.53
   Compiling imagesize v0.5.2
   Compiling xdg v2.1.0
   Compiling either v1.5.0
   Compiling leak v0.1.2
   Compiling log v0.4.1
   Compiling crossbeam-utils v0.6.1
   Compiling owning_ref v0.3.3
   Compiling unreachable v1.0.0
   Compiling textwrap v0.10.0
   Compiling synom v0.11.3
   Compiling regex-syntax v0.5.6
   Compiling rand_core v0.2.2
   Compiling arrayvec v0.4.7
   Compiling semver v0.9.0
   Compiling proc-macro2 v0.3.8
   Compiling scheduled-thread-pool v0.2.0
   Compiling cmake v0.1.33
   Compiling lru-cache v0.1.1
   Compiling humantime v1.1.1
   Compiling iovec v0.1.2
   Compiling time v0.1.40
   Compiling num_cpus v1.8.0
   Compiling net2 v0.2.32
   Compiling memchr v2.0.1
   Compiling inotify-sys v0.1.3
   Compiling filetime v0.2.3
   Compiling dirs v1.0.3
   Compiling rand v0.4.2
   Compiling memchr v1.0.2
   Compiling atty v0.2.11
   Compiling walkdir v2.1.4
   Compiling app_dirs v1.2.1
   Compiling itertools v0.7.8
   Compiling smallvec v0.6.5
   Compiling thread_local v0.3.5
   Compiling lock_api v0.1.4
   Compiling rand v0.5.5
   Compiling clang-sys v0.19.0
   Compiling log v0.3.9
   Compiling syn v0.11.11
   Compiling r2d2 v0.8.3
   Compiling bytes v0.4.10
   Compiling threadpool v1.7.1
   Compiling rustc_version v0.2.3
   Compiling crossbeam-epoch v0.6.1
   Compiling aho-corasick v0.6.4
   Compiling mio v0.6.16
   Compiling nom v3.2.1
   Compiling clap v2.32.0
   Compiling quote v0.5.2
   Compiling libsqlite3-sys v0.10.0
   Compiling tokio-executor v0.1.5
   Compiling tokio-io v0.1.10
   Compiling crossbeam-deque v0.6.2
   Compiling syn v0.13.11
   Compiling syntex_pos v0.58.1
   Compiling parking_lot_core v0.3.1
   Compiling num-traits v0.1.43
   Compiling mio-extras v2.0.5
   Compiling syntex_errors v0.58.1
   Compiling chrono v0.4.2
   Compiling cexpr v0.2.3
   Compiling parking_lot v0.6.4
   Compiling enum-primitive-derive v0.1.2
   Compiling syntex_syntax v0.58.1
   Compiling crossbeam-channel v0.3.1
   Compiling tokio-reactor v0.1.6
   Compiling inotify v0.6.1
   Compiling crossbeam v0.5.0
error[E0423]: expected function, tuple struct or tuple variant, found struct `ast::Name`
   --> /home/carlos/.cargo/registry/src/github.com-1ecc6299db9ec823/syntex_syntax-0.58.1/src/symbol.rs:146:27
    |
146 |                       name: ast::Name($index),
    |                             ^^^^^^^^^
...
165 | / declare_keywords! {
166 | |     // Invalid identifier
167 | |     (0,  Invalid,        "")
168 | |
...   |
231 | |     (56, CrateRoot, "{{root}}")
232 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

   Compiling env_logger v0.4.3
   Compiling notify v5.0.0-crossbeam (/home/carlos/dev-rust/seiri/seiri-watcher/notify)
error[E0433]: failed to resolve: use of undeclared crate or module `mpsc`
   --> notify/src/inotify.rs:470:24
    |
470 |         let (tx, rx) = mpsc::channel();
    |                        ^^^^ use of undeclared crate or module `mpsc`

error[E0433]: failed to resolve: use of undeclared crate or module `mpsc`
   --> notify/src/inotify.rs:485:24
    |
485 |         let (tx, rx) = mpsc::channel();
    |                        ^^^^ use of undeclared crate or module `mpsc`

warning: use of deprecated macro `try`: use the `?` operator instead
   --> notify/src/inotify.rs:332:24
    |
332 |         let metadata = try!(metadata(&path).map_err(Error::Io));
    |                        ^^^
    |
    = note: `#[warn(deprecated)]` on by default

warning: use of deprecated macro `try`: use the `?` operator instead
   --> notify/src/inotify.rs:343:13
    |
343 |             try!(self.add_single_watch(entry.path().to_path_buf(), is_recursive, watch_self));
    |             ^^^

warning: use of deprecated macro `try`: use the `?` operator instead
   --> notify/src/inotify.rs:395:21
    |
395 |                     try!(inotify.rm_watch(w.clone()).map_err(Error::Io));
    |                     ^^^

warning: use of deprecated macro `try`: use the `?` operator instead
   --> notify/src/inotify.rs:402:33
    |
402 | ...                   try!(inotify.rm_watch(w.clone()).map_err(Error::Io));
    |                       ^^^

warning: use of deprecated macro `try`: use the `?` operator instead
   --> notify/src/inotify.rs:420:17
    |
420 |                 try!(inotify.rm_watch(w.clone()).map_err(Error::Io));
    |                 ^^^

warning: use of deprecated macro `try`: use the `?` operator instead
   --> notify/src/inotify.rs:467:21
    |
467 |             let p = try!(env::current_dir().map_err(Error::Io));
    |                     ^^^

warning: use of deprecated macro `try`: use the `?` operator instead
   --> notify/src/inotify.rs:482:21
    |
482 |             let p = try!(env::current_dir().map_err(Error::Io));
    |                     ^^^

warning: unused import: `self`
  --> notify/src/inotify.rs:24:26
   |
24 | use crossbeam::channel::{self, Sender};
   |                          ^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused import: `Receiver`
 --> notify/src/debounce/mod.rs:9:26
  |
9 | use crossbeam::channel::{Receiver, Sender};
  |                          ^^^^^^^^

warning: unnecessary parentheses around block return value
   --> notify/src/lib.rs:461:17
    |
461 |                 (a1 == b1 && a2 == b2)
    |                 ^^^^^^^^^^^^^^^^^^^^^^ help: remove these parentheses
    |
    = note: `#[warn(unused_parens)]` on by default

warning: trait objects without an explicit `dyn` are deprecated
   --> notify/src/lib.rs:513:32
    |
513 |     fn cause(&self) -> Option<&StdError> {
    |                                ^^^^^^^^ help: use `dyn`: `dyn StdError`
    |
    = note: `#[warn(bare_trait_objects)]` on by default

warning: use of deprecated associated function `std::error::Error::description`: use the Display impl or to_string()
   --> notify/src/lib.rs:493:39
    |
493 |             Error::Io(ref err) => err.description(),
    |                                       ^^^^^^^^^^^

error: aborting due to 2 previous errors; 12 warnings emitted

For more information about this error, try `rustc --explain E0433`.
error: could not compile `notify`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: aborting due to previous error

For more information about this error, try `rustc --explain E0423`.
error: build failed
carlos@dp:~/dev-rust/seiri/seiri-watcher$ `

What do you recommend?

Carlos

**Comment by chyyran** [comment]
Unfortunately as you probably already may have noticed, seiri is rather old code that was built before many of the modern Rust idioms were solidified (like `?` !), and it was also my first foray in writing Rust. I haven't really kept it maintained since I just leave it running on my desktop, and it's been chugging along so far but I realize that that is entirely my fault.

I just encountered that error myself and it seems it's due to outdated versions of bindgen... I would update `bindgen` and `rusqlite` within Cargo.toml to their latest versions and see if it builds. 

I am a little busy this week with work but I hope to get this updated and building on newer toolchains probably before Christmas. For the time being, and I realize this is an unsatisfactory answer, I do have [Linux binaries for 0.2.7 here](https://github.com/chyyran/seiri/releases/tag/0.2.7-linux-alpha)

Otherwise, feel free to tinker around updating the rather old dependencies. I don't expect there to be too many troubles, barring any API issues.

**Comment by cv127001** [comment]
Ronny,

I think Seiri is worth the wait and I really want to start dabbling in Rust. Your project seems a good place for me to start with :)

Thanks for repeating the problem and do not hesitate if you need an extra pair of hands to run a test or two.

I'll start getting familiar with the Linux binary (thanks for pointing me to it) until I'm able to build it myself.

GoRust!

Carlos

**Comment by chyyran** [comment]
I have pushed new versions of `libkatatsuki-sys` and `katatsuki` that have updated build scripts and thus avoid the errors. 

`seiri-lib` should now be able to build if you update the `katatsuki` version to `1.0.10`. I'm working on updating the remaining dependencies, as well as the nightly requirement. Do expect an update in the coming weeks.

**Comment by cv127001** [comment]
Thank you **very** much. I'll try it out.

**Comment by cv127001** [comment]
Hi Ronny, 

I pulled the latest source. Did I do it correction with this?:

```
carlos@dp:~/Downloads$ git clone https://github.com/chyyran/seiri.git
Cloning into 'seiri'...
remote: Enumerating objects: 31, done.
remote: Counting objects: 100% (31/31), done.
remote: Compressing objects: 100% (25/25), done.
remote: Total 1949 (delta 7), reused 17 (delta 6), pack-reused 1918
Receiving objects: 100% (1949/1949), 10.81 MiB | 16.44 MiB/s, done.
Resolving deltas: 100% (1236/1236), done.
carlos@dp:~/Downloads$ 
```

If so, I got the an error:

```
carlos@dp:~/Downloads/seiri/seiri-watcher$ cargo build --release
error: failed to get `notify` as a dependency of package `seiri-watcher v0.2.0 (/home/carlos/Downloads/seiri/seiri-watcher)`

Caused by:
  failed to load source for dependency `notify`

Caused by:
  Unable to update /home/carlos/Downloads/seiri/seiri-watcher/notify

Caused by:
  failed to read `/home/carlos/Downloads/seiri/seiri-watcher/notify/Cargo.toml`

Caused by:
  No such file or directory (os error 2)
carlos@dp:~/Downloads/seiri/seiri-watcher$ 
```

Any suggestions?

Carlos

**Comment by chyyran** [comment]
Again, you need to reinit the notify submodule. I'm working on upgrading it
to use published notify versions from crates.io, but that will take a while.

As for pulling the source, the changes have been published currently in a
feature branch, but are otherwise unnecessary to have locally. You simply
need to change the `katatsuki` dependency in `seiri-lib` to point to 1.0.11

**Comment by cv127001** [comment]
Thank you Ronny. I assure you that my lack of understanding is because I'm new to Rust. I have to do some research to follow your last direction. Thank you very much! :)

**Comment by chyyran** [comment]
Will be fixed upon completion of  https://github.com/chyyran/seiri/pull/16

**Comment by chyyran** [comment]
Note that you will also need to set up neon-bindings as per https://neon-bindings.com/
JS dependencies will probably be updated by tomorrow (23) or later this week. 

I appreciate any contributions to seiri, but if you're new to Rust it may not be the easiest project to get started with, not necessarily because of the Rust part which is fairly simple, but because the build steps just so involved. seiri interacts with three or four different programming languages (Rust, C++, C, JavaScript) that all have to talk to each other, so the build process is very involved and not very easy to get started with. That's just the nature of having so many different moving part unfortunately. It can definitely be improved, but since I usually update this once a year, the motivation isn't really there.

**Comment by chyyran** [comment]
Fixed by #16
